### PR TITLE
Implements Command Watch Manager and Adds GET.WATCH Command Support

### PR DESCRIPTION
**Pull request by JyotinderSingh**
# Implements Command Watch Manager and Adds GET.WATCH Command Support

This PR introduces a new feature that allows clients to watch for changes on specific keys through command watches. 

The primary changes include:

- **Watch Manager Implementation**: Added a new `watchmanager` package that handles watch subscriptions and notifications for commands affecting keys. The watch manager maintains mappings of subscriptions and notifies clients when watched keys are affected by specific commands.

- **GET.WATCH Command Support**: Implemented support for the `GET.WATCH` command, allowing clients to watch a key and receive real-time updates when the key's value changes. The server sends push notifications to the client whenever the watched key is updated.

- **Server and Worker Modifications**:
  - Modified the `Server` and `Worker` components to integrate with the watch manager.
  - Added a new channel `cmdWatchChan` to communicate command watch events.
  - Workers can now handle watch subscriptions and execute commands to notify clients.

- **Client IO Adjustments**: Updated the `clientio` package to handle push responses for command watches. The `CreatePushResponse` function now includes the key as an identifier in the push response.

- **Command Modifications**:
  - Added `GetFingerprint` and `GetKey` methods to the `DiceDBCmd` struct for identifying commands in the watch manager.
  - Updated command metadata to include a new command type `Watch`.

- **Store Updates**:
  - Modified the `Store` to notify the watch manager about command events via the new `cmdWatchChan`.
  - The store now distinguishes between query watch events and command watch events.

- **Integration Tests**:
  - Added new tests for the `GET.WATCH` command in `integration_tests/commands/resp/getwatch_test.go`.
  - Adjusted existing tests to accommodate the new changes, including renaming functions and updating assertion messages.

### CODE REVIEW

**Review by lucifercr07 — CHANGES_REQUESTED** [review]
Thanks a lot for changes, LGTM. Minor clarification comments.

File: internal/watchmanager/watch_manager.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+func (m *Manager) listenForEvents(ctx context.Context, cmdWatchChan chan dstore.CmdWatchEvent) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case sub := <-CmdWatchSubscriptionChan:
+			if sub.Subscribe {
+				m.handleSubscription(sub)
+			} else {
+				m.handleUnsubscription(sub)
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply internal/watchmanager/watch_manager.go:70]: Yes, that will be part of the next PR

File: internal/store/store.go [hunk 2]
```diff
@@ -149,9 +156,12 @@ func (store *Store) putHelper(k string, obj *object.Obj, opts ...PutOption) {
 	}
 	store.store.Put(k, obj)
 
-	if store.watchChan != nil {
+	if store.queryWatchChan != nil {
 		store.notifyQueryManager(k, Set, *obj)
 	}
+	if store.cmdWatchChan != nil {
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply internal/store/store.go:162]: That will be covered once we have something like response fingerprinting implemented. We will hold onto fingerprints of the latest responses sent to the client and only send new responses if they are different from what was sent before.
That should be done as a part of subsequent PRs.

File: internal/worker/worker.go [hunk 3]
```diff
@@ -295,23 +353,39 @@ func (w *BaseWorker) gather(ctx context.Context, c string, numCmds int, ct CmdTy
...
 		if err != nil {
 			w.logger.Debug("Error sending response to client", slog.String("workerID", w.id), slog.Any("error", err))
 			return err
 		}
 	}
+	return nil
+}
+
+// handlePushResponse handles the response for push commands, i.e., responses for which are pushed from the server to the client.
+func (w *BaseWorker) handlePushResponse(ctx context.Context, cmdName, pushResponseKey string, evalResp []eval.EvalResponse) error {
```
Comment (soumya-codes) [inline internal/worker/worker.go:375]: Do we need special handling for watch notification for handling response?
Can we use the the regular response flow instead?

**Review by soumya-codes — COMMENTED** [review]
The changes look great! Left a few minor comments.

File: internal/watchmanager/watch_manager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply internal/watchmanager/watch_manager.go:70]: #998

File: internal/watchmanager/watch_manager.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	Manager struct {
+		querySubscriptionMap map[string]map[uint32]struct{}              // querySubscriptionMap is a map of Key -> [fingerprint1, fingerprint2, ...]
+		tcpSubscriptionMap   map[uint32]map[chan *cmd.DiceDBCmd]struct{} // tcpSubscriptionMap is a map of fingerprint -> [client1Chan, client2Chan, ...]
+		fingerprintCmdMap    map[uint32]*cmd.DiceDBCmd                   // fingerprintCmdMap is a map of fingerprint -> DiceDBCmd
+		logger               *slog.Logger
+	}
+)
+
+var (
+	CmdWatchSubscriptionChan chan WatchSubscription
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply internal/watchmanager/watch_manager.go:29]: #1000

File: internal/worker/worker.go [hunk 5]
```diff
@@ -52,6 +55,7 @@ func NewWorker(wid string, respChan chan *ops.StoreResponse,
 		respChan:        respChan,
 		logger:          logger,
 		Session:         auth.NewSession(),
+		adhocReqChan:    make(chan *cmd.DiceDBCmd, 20), // assuming we wouldn't have more than 20 adhoc requests being sent at a time.
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply internal/worker/worker.go:58]: Good point. Need to think how this should be handled. I'll create a new issue for this: #1001

**Pull request merged**